packages/context/src/__tests__/unit/context-config.unit.ts

```diff
@@ -0,0 +1,131 @@
+  BindingKey,
+  ConfigurationResolver,
+  Context,
+  DefaultConfigurationResolver,
+  ResolutionOptions,
+  ValueOrPromise,
+} from '../..';
+
+describe('Context binding configuration', () => {
+  /**
+   * Create a subclass of context so that we can access parents and registry
+   * for assertions
+   */
+  class TestContext extends Context {
+    public configResolver: ConfigurationResolver;
+  }
+
+  let ctx: TestContext;
+  beforeEach(createContext);
+
+  describe('configure()', () => {
+    it('configures options for a binding before it is bound', () => {
+      const bindingForConfig = ctx.configure('foo').to({x: 1});
+      expect(bindingForConfig.key).to.equal(
+        BindingKey.buildKeyForConfig('foo'),
+      );
+      expect(bindingForConfig.tagMap).to.eql({config: 'foo'});
+    });
+
+    it('configures options for a binding after it is bound', () => {
+      ctx.bind('foo').to('bar');
+      const bindingForConfig = ctx.configure('foo').to({x: 1});
+      expect(bindingForConfig.key).to.equal(
+        BindingKey.buildKeyForConfig('foo'),
+      );
+      expect(bindingForConfig.tagMap).to.eql({config: 'foo'});
+    });
+  });
+
+  describe('getConfig()', () => {
+    it('gets config for a binding', async () => {
+      ctx.configure('foo').toDynamicValue(() => Promise.resolve({x: 1}));
+      expect(await ctx.getConfig('foo')).to.eql({x: 1});
+    });
+
+    it('gets local config for a binding', async () => {
+      ctx
+        .configure('foo')
+        .toDynamicValue(() => Promise.resolve({a: {x: 0, y: 0}}));
+      ctx.configure('foo.a').toDynamicValue(() => Promise.resolve({x: 1}));
+      expect(await ctx.getConfig<number>('foo.a', 'x')).to.eql(1);
+      expect(await ctx.getConfig<number>('foo.a', 'y')).to.be.undefined();
+    });
+
+    it('defaults optional to true for config resolution', async () => {
+      // `servers.rest` does not exist yet
+      let server1port = await ctx.getConfig<number>('servers.rest', 'port');
+      expect(server1port).to.be.undefined();
+
+      // Now add `servers.rest`
+      ctx.configure('servers.rest').to({port: 3000});
+      server1port = await ctx.getConfig<number>('servers.rest', 'port');
+      expect(server1port).to.eql(3000);
+    });
+
+    it('throws error if a required config cannot be resolved', async () => {
+      await expect(
+        ctx.getConfig('servers.rest', 'host', {
+          optional: false,
+        }),
+      ).to.be.rejectedWith(/The key 'servers\.rest\:\$config' is not bound/);
+    });
+  });
+
+  describe('getConfigSync()', () => {
+    it('gets config for a binding', () => {
+      ctx.configure('foo').to({x: 1});
+      expect(ctx.getConfigSync('foo')).to.eql({x: 1});
+    });
+
+    it('throws a helpful error when the config is async', () => {
+      ctx.configure('foo').toDynamicValue(() => Promise.resolve('bar'));
+      expect(() => ctx.getConfigSync('foo')).to.throw(
+        /Cannot get config\[\] for foo synchronously: the value is a promise/,
```

Comment:
Where is the `[]` coming from? Is it because there is no property path specified? Can we leave it out in such case?

Please consider adding a test for `ctx.getConfigSync('key', 'path')` too.